src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobWorker.cs

```diff
@@ -0,0 +1,78 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Core.Configs;
+using Microsoft.Health.Fhir.Core.Features.Operations.Export.Models;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Export
+{
+    /// <summary>
+    /// The worker responsible for running the export job tasks.
+    /// </summary>
+    public class ExportJobWorker
+    {
+        private readonly IFhirOperationDataStore _fhirOperationDataStore;
+        private readonly ExportJobConfiguration _exportJobConfiguration;
+        private readonly IExportJobTaskFactory _exportJobTaskFactory;
+        private readonly ILogger _logger;
+
+        public ExportJobWorker(IFhirOperationDataStore fhirOperationDataStore, IOptions<ExportJobConfiguration> exportJobConfiguration, IExportJobTaskFactory exportJobTaskFactory, ILogger<ExportJobWorker> logger)
+        {
+            EnsureArg.IsNotNull(fhirOperationDataStore, nameof(fhirOperationDataStore));
+            EnsureArg.IsNotNull(exportJobConfiguration?.Value, nameof(exportJobConfiguration));
+            EnsureArg.IsNotNull(exportJobTaskFactory, nameof(exportJobTaskFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _fhirOperationDataStore = fhirOperationDataStore;
+            _exportJobConfiguration = exportJobConfiguration.Value;
+            _exportJobTaskFactory = exportJobTaskFactory;
+            _logger = logger;
+        }
+
+        public async Task ExecuteAsync(CancellationToken cancellationToken)
+        {
+            var runningTasks = new List<Task>();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    // Remove all completed tasks.
+                    runningTasks.RemoveAll(task => task.IsCompleted);
+
+                    // Get list of available jobs.
+                    if (runningTasks.Count < _exportJobConfiguration.MaximumNumberOfConcurrentJobsAllowed)
+                    {
+                        IReadOnlyCollection<ExportJobOutcome> jobs = await _fhirOperationDataStore.GetAvailableExportJobsAsync(
+                            _exportJobConfiguration.MaximumNumberOfConcurrentJobsAllowed,
```

Comment:
We should pass in `_exportJobConfiguration.MaximumNumberOfConcurrentJobsAllowed - runningTasks.Count` here. Else we can end up adding more number of jobs than the `MaximumNumberOfConcurrentJobsAllowed` number.

Reply:
We should pass in `_exportJobConfiguration.MaximumNumberOfConcurrentJobsAllowed - runningTasks.Count` here. Else we can end up adding more number of jobs than the `MaximumNumberOfConcurrentJobsAllowed` number.

Reply:
I just noticed that the stored procedure takes care of this by querying for the number of running jobs before returning the available jobs. Is there a reason we don't want to have that logic here  instead of in the stored procedure?

Reply:
If there are multiple instances of worker running at the same time, you need something (BE in this case) to centrally control that. It would be harder (I think) to enforce global rule from distributed workers.